group Cells by type in cells_dict

https://github.com/nschloe/meshio/blob/ac2ca8428d1b9a2eed9a220f043cf16f4b99b425/meshio/_mesh.py#L120-L125

fails if there is more than one item of `Mesh.cells` that has the same type.

This case does arise in practice, e.g. if the `Mesh` is read from MSH 4.1 generated by Gmsh, which makes a new item for each of the MSH 4.1 `$Entities`; see nschloe/pygmsh#301.

It might be reasonable to group the items of `.cells` by type, concatenating the `Cells.data`.

A simple example.  Starting with 
* [disk.geo](https://github.com/nschloe/meshio/files/4189278/disk.geo.txt) (adapted from nschloe/pygmsh#301, not using pygmsh)
```
Point(1) = {0, 0, 0};
Point(2) = {1, 0, 0};
Point(3) = {-0.5, Sqrt(3)/2, 0};
Point(4) = {-0.5, -Sqrt(3)/2, 0};

Circle(1) = {2, 1, 3};
Circle(2) = {3, 1, 4};
Circle(3) = {4, 1, 2};
Line Loop(1) = {1, 2, 3};
Plane Surface(1) = {1};
Physical Line("circle") = {1, 2, 3};
Physical Surface("disk") = {1};
```

run
```shell
gmsh -2 disk.geo
```
to generate (here using Gmsh 4.5.1 from `pip install gmsh-sdk`)
* [disk.msh](https://github.com/nschloe/meshio/files/4189280/disk.msh.txt)

Then `meshio.read` gives a `Mesh` which prints as
```
<meshio mesh object>
  Number of points: 118
  Number of cells:
    line: 10
    line: 10
    line: 10
    triangle: 204
  Cell data: gmsh:physical
```
and for which `.cells_dict` raises:
```
AssertionError: More than one block of the same type. Cannot create dictionary.
```

--- a/meshio/_mesh.py
+++ b/meshio/_mesh.py
@@ -129,6 +129,20 @@ class Mesh:
             cells_dict[key] = numpy.concatenate(value)
         return cells_dict
 
+    @property
+    def cell_data_dict(self):
+        cell_data_dict = {}
+        for key, value_list in self.cell_data.items():
+            cell_data_dict[key] = {}
+            for value, (cell_type, _) in zip(value_list, self.cells):
+                if cell_type not in cell_data_dict[key]:
+                    cell_data_dict[key][cell_type] = []
+                cell_data_dict[key][cell_type].append(value)
+
+            for cell_type, val in cell_data_dict[key].items():
+                cell_data_dict[key][cell_type] = numpy.concatenate(val)
+        return cell_data_dict
+
     @classmethod
     def read(cls, path_or_buf, file_format=None):
         # avoid circular import

--- a/meshio/_mesh.py
+++ b/meshio/_mesh.py
@@ -119,10 +119,15 @@ class Mesh:
 
     @property
     def cells_dict(self):
-        assert len(self.cells) == len(
-            numpy.unique([c.type for c in self.cells])
-        ), "More than one block of the same type. Cannot create dictionary."
-        return dict(self.cells)
+        cells_dict = {}
+        for cell_type, data in self.cells:
+            if cell_type not in cells_dict:
+                cells_dict[cell_type] = []
+            cells_dict[cell_type].append(data)
+        # concatenate
+        for key, value in cells_dict.items():
+            cells_dict[key] = numpy.concatenate(value)
+        return cells_dict
 
     @classmethod
     def read(cls, path_or_buf, file_format=None):

--- a/meshio/stl/_stl.py
+++ b/meshio/stl/_stl.py
@@ -148,7 +148,7 @@ def write(filename, mesh, binary=False):
     if not any(c.type == "triangle" for c in mesh.cells):
         raise WriteError(
             "STL can only write triangle cells (not {}).".format(
-                ", ".join(list(mesh.cells.keys()))
+                ", ".join(c.type for c in mesh.cells)
             )
         )
     if len(mesh.cells) > 1:

--- a/test/test_mesh.py
+++ b/test/test_mesh.py
@@ -27,6 +27,8 @@ def test_cells_dict():
             ("triangle", numpy.array([[0, 1, 2]])),
             ("triangle", numpy.array([[0, 2, 3]])),
         ],
+        cell_data={"a": [[0.5], [1.3]]},
     )
     assert len(mesh.cells_dict) == 1
     assert numpy.array_equal(mesh.cells_dict["triangle"], [[0, 1, 2], [0, 2, 3]])
+    assert numpy.array_equal(mesh.cell_data_dict["a"]["triangle"], [0.5, 1.3])

--- a/test/test_mesh.py
+++ b/test/test_mesh.py
@@ -3,6 +3,7 @@ import copy
 import numpy
 
 import helpers
+import meshio
 
 
 def test_print_prune():
@@ -13,5 +14,19 @@ def test_print_prune():
 
 def test_cells_dict():
     mesh = copy.deepcopy(helpers.tri_mesh)
+    assert len(mesh.cells_dict) == 1
+    assert numpy.array_equal(mesh.cells_dict["triangle"], [[0, 1, 2], [0, 2, 3]])
+
+    # two cells groups
+    mesh = meshio.Mesh(
+        numpy.array(
+            [[0.0, 0.0, 0.0], [1.0, 0.0, 0.0], [1.0, 1.0, 0.0], [0.0, 1.0, 0.0]]
+        )
+        / 3,
+        [
+            ("triangle", numpy.array([[0, 1, 2]])),
+            ("triangle", numpy.array([[0, 2, 3]])),
+        ],
+    )
     assert len(mesh.cells_dict) == 1
     assert numpy.array_equal(mesh.cells_dict["triangle"], [[0, 1, 2], [0, 2, 3]])

